Add new Lock core component

kbickar opened the pull request:
# What does this implement/fix? 
Home Assistant has "[lock](https://www.home-assistant.io/integrations/lock/)" domain which has some differences from other domains like "switch" or "cover" such as extra security restrictions.

For smart locks made with an ESPx, it is useful to have this as a separate component that is then available natively in HomeAssistant

Added a template lock and a generic output lock to go along with it

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/612

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** 
esphome/esphome-docs#1867
esphome/aioesphomeapi#158
home-assistant/core#65280

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [X] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml

output:
  - platform: gpio
    pin: D8
    inverted: true
    id: lockpin
    
lock:
 - platform: output
   output: lockpin
   name: Test Lock

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

kbickar committed:
Created Lock Core Component

kbickar committed:
Create Template Lock Component

kbickar committed:
Create Output Lock Component

kbickar committed:
Remove Toggle

kbickar committed:
Add lock to APIs

kbickar committed:
Bug Fixes

kbickar committed:
Handy dev lib

kbickar committed:
Add Open button

kbickar committed:
Minor Fixes

kbickar committed:
Minor fixes

kbickar committed:
Formatting

kbickar committed:
Guess it was just barely long enough

kbickar committed:
Added lock to custom linter test exceptions

kbickar committed:
Add core codeowner

kbickar committed:
Align API protocol

kbickar committed:
Add tests

kbickar committed:
Fix Spaces

kbickar committed:
Fix spaces

jesserockz commented on docker/Dockerfile:
```diff
@@ -20,30 +20,31 @@ RUN \
     apt-get update \
     # Use pinned versions so that we get updates with build caching
     && apt-get install -y --no-install-recommends \
-        python3=3.9.2-3 \
```
Please remove the Dockerfile changes from the PR

kbickar committed:
Revert dockerfile change

kbickar committed:
Add support for extended lock modes

kbickar committed:
Fix some tests

kbickar committed:
Code cleanup

kbickar committed:
API updates

kbickar commented:
Just a note on the new modes I added -
HomeAssistant internally has support for lock states other than LOCKED/UNLOCKED, however the MQTT component does not

oxan commented on esphome/components/lock/__init__.py:
```diff
@@ -0,0 +1,107 @@
+LOCK_SCHEMA = cv.ENTITY_BASE_SCHEMA.extend(cv.MQTT_COMMAND_COMPONENT_SCHEMA).extend(
+    {
+        cv.OnlyWith(CONF_MQTT_ID, "mqtt"): cv.declare_id(mqtt.MQTTLockComponent),
+        cv.Optional(CONF_INVERTED): cv.boolean,
```
This doesn't seem to be used.

oxan commented on esphome/components/lock/__init__.py:
```diff
@@ -0,0 +1,107 @@
+)
+
+
+@automation.register_action("lock.unlock", UnlockAction, LOCK_ACTION_SCHEMA)
+@automation.register_action("lock.lock", LockAction, LOCK_ACTION_SCHEMA)
+@automation.register_action("lock.open", OpenAction, LOCK_ACTION_SCHEMA)
```
Implementation missing.

kbickar replied:
Are the implementations not in automation.h?  They seem to work

oxan replied:
The Python function to convert these to code is missing. Right now it uses `lock_is_on_to_code`, since that's the function these decorators are on, but that one doesn't work, as it gets an extra constructor parameter.

oxan commented on esphome/components/lock/automation.cpp:
```diff
@@ -0,0 +1,10 @@
+static const char *const TAG = "lock.automation";
+
+}  // namespace lock
+}  // namespace esphome
```
This file can be left out.

kbickar replied:
The compiler is requiring it

oxan replied:
What do you mean?

kbickar replied:
Disregard, was just still listed in the git cache

oxan commented on esphome/components/lock/automation.h:
```diff
@@ -0,0 +1,90 @@
+      return this->parent_->state == LockState::LOCK_STATE_LOCKED;
+    if (!this->state_)
+      return this->parent_->state == LockState::LOCK_STATE_UNLOCKED;
+    return false;
```
This is dead code, a bool has just two possible values :) I would either drop the second if (and maybe invert the order then), or do something like this:
```c++
    auto check_state = this->state_ ? LockState::LOCK_STATE_LOCKED : LockState::LOCK_STATE_UNLOCKED;
    return this->parent_->state == check_state;
```

kbickar replied:
That's a bit clearer of a way to do it

oxan commented on esphome/components/lock/lock.cpp:
```diff
@@ -0,0 +1,79 @@
+Lock::Lock() : Lock("") {}
+
+void Lock::lock() {
+  ESP_LOGD(TAG, "'%s' LOCKING.", this->get_name().c_str());
```
```suggestion
  ESP_LOGD(TAG, "'%s' Locking.", this->get_name().c_str());
```

oxan commented on esphome/components/lock/lock.cpp:
```diff
@@ -0,0 +1,79 @@
+  this->write_state(LOCK_STATE_LOCKED);
+}
+void Lock::unlock() {
+  ESP_LOGD(TAG, "'%s' UNLOCKING.", this->get_name().c_str());
```
```suggestion
  ESP_LOGD(TAG, "'%s' Unlocking.", this->get_name().c_str());
```

oxan commented on esphome/components/lock/lock.cpp:
```diff
@@ -0,0 +1,79 @@
+    ESP_LOGD(TAG, "'%s' Opening.", this->get_name().c_str());
+    this->open_latch();
+  } else {
+    ESP_LOGD(TAG, "'%s' Does not support Open.", this->get_name().c_str());
```
```suggestion
    ESP_LOGW(TAG, "'%s' Does not support Open.", this->get_name().c_str());
```

oxan commented on esphome/components/lock/lock.cpp:
```diff
@@ -0,0 +1,79 @@
+bool Lock::assumed_state() { return false; }
+
+void Lock::add_on_state_callback(std::function<void()> &&callback) { this->state_callback_.add(std::move(callback)); }
+uint32_t Lock::hash_base() { return 3129890955UL; }
```
This should be a different number from the one used for switches.

oxan commented on esphome/components/lock/lock.h:
```diff
@@ -0,0 +1,116 @@
+
+  /** Publish a state to the front-end from the back-end.
+   *
+   * The input value is inverted if applicable. Then the internal value member is set and
```
Inapplicable comment.

oxan commented on esphome/components/lock/lock.h:
```diff
@@ -0,0 +1,116 @@
+  /// The current reported state of the lock.
+  LockState state{LOCK_STATE_NONE};
+
+  bool supports_open{false};
+  bool requires_code{false};
```
We don't implement these kind of static properties with a public member, since those are writable for everyone. Instead, we use a traits object, which also allows us to more cleanly add additional properties later on. See e.g. how the `number` component handles the min/max value.

oxan commented on esphome/components/lock/lock.h:
```diff
@@ -0,0 +1,116 @@
+   *
+   * Defaults to false.
+   */
+  virtual bool assumed_state();
```
This should also be in the traits.

oxan commented on esphome/components/mqtt/mqtt_lock.cpp:
```diff
@@ -0,0 +1,55 @@
+    else {
+      ESP_LOGW(TAG, "'%s': Received unknown status payload: %s", this->friendly_name().c_str(), payload.c_str());
+      this->status_momentary_warning("state", 5000);
+    }
```
Braces for all or none of the branches, please (I should add a CI check to catch that).

oxan commented on esphome/core/controller.h:
```diff
@@ -70,6 +73,9 @@ class Controller {
 #endif
+#ifdef USE_LOCK
+  virtual void on_lock_update(lock::Lock *obj){};
+#endif
```
I'd like a state parameter here for consistency.

kbickar replied:
It's a bit of a pain to use the state there since it's not a native type.  It's consistent with fan, light, cover, and climate.  Seems to be a bit of a redundant parameter as it just uses the state of the object for the others

oxan replied:
I don't see why having `LockState` here would be problematic, but good point about the others also missing it.

oxan commented on esphome/components/lock/lock.cpp:
```diff
@@ -0,0 +1,79 @@
+    default:
+      return "UNKNOWN";
+  }
+}
```
The reason we have the `LogString` machinery is so that strings that are only used for logging can be stored in flash instead of in valuable RAM. However, if we then include a second copy of those strings in RAM, that didn't gain us anything. Initially I suggested to just drop the non-LogString version, but I see it's being used for MQTT. Eventually I want to create a proper `FlashString` class that allows conversion to `std::string` for that, but for now just drop the `LogString` version.

Please change the return type to `const char *` though, creating a `std::string` here is unnecessary most of the time.

oxan commented on docker/Dockerfile:
```diff
@@ -129,23 +129,23 @@ LABEL \
-        /var/lib/apt/lists/*
+    /tmp/* \
+    /var/{cache,log}/* \
+    /var/lib/apt/lists/*
```
Please remove the unrelated whitespace changes to this file.

oxan commented on esphome/components/api/api.proto:
```diff
@@ -977,3 +978,57 @@ message ButtonCommandRequest {
+  option (ifdef) = "USE_LOCK";
+  option (no_delay) = true;
+  fixed32 key = 1;
+  LockState  state = 2;
```
```suggestion
  LockState state = 2;
```

oxan commented on esphome/components/api/api.proto:
```diff
@@ -977,3 +978,57 @@ message ButtonCommandRequest {
+  EntityCategory entity_category = 7;
+  bool assumed_state = 8;
+
+  bool supports_open = 9;
+  bool requires_code = 10;
+  string code_format = 11;
```
These don't seem to be set anywhere?

kbickar replied:
Added `supports_open` to api_connection.cpp. The other two are there to match the API that @jesserockz used in esphome/aioesphomeapi#158

oxan replied:
Ah, okay. As long as they aren't implemented, they can be removed imo, but let's at least add a comment that they aren't used yet.

oxan reviewed (changes requested):
Thanks for working on this! 

Don't let my comments discourage you, overall it looks fairly good, but I like to get new platforms such as this right the first time, since changing them later is often painful for both us and our users.

kbickar committed:
Updates from review

kbickar committed:
Merge branch 'esphome:dev' into lock

kbickar committed:
Changes for new Arduino lib

kbickar committed:
Update function def

kbickar committed:
Update protocol ids

kbickar committed:
Updates from review

oxan commented on esphome/components/lock/__init__.py:
```diff
@@ -0,0 +1,108 @@
+LockLockTrigger = lock_ns.class_("LockLockTrigger", automation.Trigger.template())
+LockUnlockTrigger = lock_ns.class_("LockUnlockTrigger", automation.Trigger.template())
+
+icon = cv.icon
```
This is unused.

oxan commented on esphome/components/lock/__init__.py:
```diff
@@ -0,0 +1,108 @@
+    for conf in config.get(CONF_ON_UNLOCK, []):
+        trigger = cg.new_Pvariable(conf[CONF_TRIGGER_ID], var)
+        await automation.build_automation(trigger, [], conf)
+    for conf in config.get(CONF_ON_OPEN, []):
```
The config option for this doesn't actually seem to exist.

oxan commented on esphome/components/lock/lock.h:
```diff
@@ -0,0 +1,129 @@
+   * Assistant because the real state is unknown.
+   *
+   * Defaults to false.
+   */
```
Wild comment.

oxan commented on esphome/components/template/lock/__init__.py:
```diff
@@ -0,0 +1,93 @@
+        CONF_LOCK_ACTION not in config or CONF_UNLOCK_ACTION not in config
+    ):
+        raise cv.Invalid(
+            "Either optimistic mode must be enabled, or turn_on_action and turn_off_action must be set, "
```
```suggestion
            "Either optimistic mode must be enabled, or lock_action and unlock_action must be set, "
```

oxan reviewed (changes requested):
Looks pretty good overall, just some minor mistakes I spotted.

kbickar committed:
Minor update from review

kbickar committed:
Remove unused import

jchang10 commented:
I hope the powers that be can get this thru. I am surprised that "lock" is not supported by esphome at this time already. Great work so far! Let's hope this gets done soon!

jesserockz commented on esphome/components/api/api_connection.cpp:
```diff
@@ -698,6 +698,49 @@ void APIConnection::button_command(const ButtonCommandRequest &msg) {
+
+  LockStateResponse resp{};
+  resp.key = a_lock->get_object_id_hash();
+  resp.state = (enums::LockState) state;
```
To be consistent with other code in here
```suggestion
  resp.state = static_cast<enums::LockState>(state);
```

jesserockz reviewed (changes requested):
I feel like even though it will be very simple now, I would like there to be a `LockCall` similar to the other components, which is called from actions, api/mqtt commands etc and does validation (future for when `code` is supported)

The rest of the code looks good to me.

kbickar committed:
Add LockCall class, remove write_state() for control()

kbickar committed:
Fix merge conflicts

kbickar committed:
Merge upstream

kbickar committed:
Merge branch 'dev' into lock

kbickar committed:
Updated linting

kbickar commented:
> I feel like even though it will be very simple now, I would like there to be a `LockCall` similar to the other components, which is called from actions, api/mqtt commands etc and does validation (future for when `code` is supported)

@jesserockz , I updated it with LockCall and changed the `write_state()` to `control()` since it doesn't really make sense to require both

jesserockz commented on esphome/components/template/lock/__init__.py:
```diff
@@ -0,0 +1,93 @@
+    cv.Schema(
+        {
+            cv.Required(CONF_ID): cv.use_id(lock.Lock),
+            cv.Required(CONF_STATE): cv.templatable(LockState),
```
```suggestion
            cv.Required(CONF_STATE): cv.templatable(validate_lock_state),
```

jesserockz commented on esphome/components/template/lock/__init__.py:
```diff
@@ -0,0 +1,93 @@
+TemplateLock = template_ns.class_("TemplateLock", lock.Lock, cg.Component)
+
+LockState = template_ns.enum("LockState")
+
```
```suggestion

LOCK_STATES = {
    "LOCKED": LockState.LOCK_STATE_LOCKED,
    "UNLOCKED": LockState.LOCK_STATE_UNLOCKED,
    "JAMMED": LockState.LOCK_STATE_JAMMED,
    "LOCKING": LockState.LOCK_STATE_LOCKING,
    "UNLOCKING": LockState.LOCK_STATE_UNLOCKING,
}

validate_lock_state = cv.enum(LOCK_STATES, upper=True)
```

jesserockz reviewed (changes requested):
Actually one thing I missed and noticed while reading the docs PR

kbickar committed:
Update esphome/components/template/lock/__init__.py

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

kbickar committed:
Update esphome/components/template/lock/__init__.py

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

kbickar committed:
Formatting

OttoWinter commented on esphome/components/lock/lock.cpp:
```diff
@@ -0,0 +1,111 @@
+LockCall Lock::make_call() { return LockCall(this); }
+
+void Lock::lock() {
+  ESP_LOGD(TAG, "'%s' Locking.", this->get_name().c_str());
```
This debug log should be in LockCall::perform instead, as otherwise we'd have inconsistent logging if the call executed by manually creating a LockCall

OttoWinter commented on esphome/components/template/lock/template_lock.h:
```diff
@@ -0,0 +1,40 @@
+namespace esphome {
+namespace template_ {
+
+using namespace esphome::lock;
```
We shouldn't have using namespace in header files, as now the lock namespace gets included in the template_ namespace for all files that include this file, but it really is more of an implementation detail.

kbickar replied:
It allows templates with lambda statements like `return LOCK_STATE_LOCKED;` which is a little more user friendly.  I wasn't able to see how to include that namespace there otherwise

jesserockz replied:
All platform components add their namespace with `cg.add_global(lock_ns.using)` in the `to_code` function.

https://github.com/esphome/esphome/pull/2958/files#diff-d195c5a59ace2a6f8e8ecfa2666f71deda7c37b06330e27d6259f5602ec4b53aR99-R102

kbickar replied:
If I remove that line then it doesn't seem to work though

jesserockz replied:
Ah, move it to `template_lock.cpp` (See `template_cover.cpp`)

kbickar committed:
Removed extra logging

kbickar committed:
Remove namespace from template header

kbickar committed:
Added back a little namespace

jesserockz commented on esphome/components/template/lock/template_lock.cpp:
```diff
@@ -0,0 +1,62 @@
+  this->prev_trigger_ = this->open_trigger_;
+  this->open_trigger_->trigger();
+
+  if (this->optimistic_)
+    this->publish_state(LOCK_STATE_UNLOCKED);
```
I don't think triggering an open should publish the state? 
If the door is locked and you try open it, it doesn't try to unlock, and there is no logic preventing the lock from trying to open while locked. So I think just remove these lines. There doesn't need to be feedback for the `opening` of a lock as it is just a side function. 
```suggestion
```

kbickar replied:
That part I just added to be consistent with the HA MQTT component: https://www.home-assistant.io/integrations/lock.mqtt/

> If the lock is in optimistic mode, it will change states to unlocked when handling the open command.

jesserockz reviewed (changes requested):
One last change sorry =)

kbickar committed:
Update esphome/components/template/lock/template_lock.cpp

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

kbickar committed:
Remove extra line
